### PR TITLE
Work/logic refactoring

### PR DESCRIPTION
**Pull request by Qrivi**
Application-logic-wise these additions are _imo_ really nice and solid, UI-wise this branch is a minor step back in its current state. Still a work in progress, but because this branch is getting huge, perhaps it's "ready enough" to become the new `develop` so we can start working on smaller stuff branched off of the progress made here. 

Extra application logic and refactoring into a new and maintainable architecture is done, but not all UI widgets were reviewed/put back. The code for these widgets is in `lib/views/backup`. Perhaps the biggest step back are the swipeable cards missing. I found the existing implementation extremely messy and I have started working on my own on the `work/swipeable-card` branch. My implementation will have 1 contoller to manage card state which should result in better performance and cleaner code. It will also be a `StateNotifierProvider` so the top card/photos will be easily accessible accros views (like on the swipe page and on the animal detail view without needing to pass state as arguments).

In the end up to @Pekket to decide if we can merge and start working on this codebase, or if we wait a bit longer till more UI is ported. It is also an option to bring back the old TCards until  `work/swipeable-card` is ready but tbh imo that would be a waste of time. I would have more time to work on my implementation if this branch becomes the new `develop` and I don't need to backport existing UI widgets all by myself. :-)

Please review/add comments if unclear or bad code.

### Notable changes:

- `ApiClient` now handles processing of response headers and response body, including exception handling and QOL features like timeouts or exceptions when offline. It's wrapped in a Provider to be accessible to the repositories.
- `DeviceData` is a new provider with methods to get the user's current locale, current location and current "user agent". That last one is much improved and will make for a nice string like "Xiaomi Redmi Note 7 (Android 10)" or "Apple iPhone 12 Pro (iOS 15.1)".
- Codebase now has "repositories" that 1:1 map every available API endpoint, and models to map these JSON response bodies onto.
- Codebase also has "controllers" which are `StateNotifierProviders` to consume in the UI, that use/combine data from the aforementioned repositories and make them available in clean state models.

### CODE REVIEW

**Comment by Qrivi** [comment]
The diff is so big it's probably easier to review the changes by checking out this branch and going over everything in the `lib` folder except `lib/views`. The latter contains old UI code, but all non-UI code was rewritten.

Please leave feedback though. No way there's nothing I could have written better. :D All feedback welcome and appreciated.

**Comment by Qrivi** [comment]
Quality checks failing because CI is using Dart <2.15. 😩

- [x] Update GitHub Actions workflow to use latest Flutter/Dart SDKs.

File: lib/providers/api_client.dart [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    return _parseResponse(
+        () async => await super.delete(url, headers: await _authIfNeeded(url, headers), body: body, encoding: encoding).timeout(RescadoConstants.timeout),
+      );
+  }
+
+  // If the url is not a public url, add the authorization header.
+  Future<Map<String, String>?> _authIfNeeded(Uri url, Map<String, String>? headers) async {
+    if (!url.path.contains('/auth/')) {
+      headers ??= {};
+      headers[HttpHeaders.authorizationHeader] = 'Bearer ${await RescadoStorage.getToken()}';
```
Comment (Pekket) [inline lib/providers/api_client.dart]: Is `SharedPreferences.getInstance()` not a heavy operation every time /auth is called?

Comment (Qrivi) [reply lib/providers/api_client.dart]:
```dart
  /// Loads and parses the [SharedPreferences] for this app from disk.
  ///
  /// Because this is reading from disk, it shouldn't be awaited in
  /// performance-sensitive blocks.
  static Future<SharedPreferences> getInstance() async {
    if (_completer == null) {
      final Completer<SharedPreferences> completer =
          Completer<SharedPreferences>();
      try {
        final Map<String, Object> preferencesMap =
            await _getSharedPreferencesMap();
        completer.complete(SharedPreferences._(preferencesMap));
      } on Exception catch (e) {
        // If there's an error, explicitly return the future with an error.
        // then set the completer to null so we can retry.
        completer.completeError(e);
        final Future<SharedPreferences> sharedPrefsFuture = completer.future;
        _completer = null;
        return sharedPrefsFuture;
      }
      _completer = completer;
    }
    return _completer!.future;
  }
  ```

tl;dr a whole lot of checks the first time but once done, it keeps a reference to itself and just returns that, which is instant. aka singleton. 🙂

Good remark though. I am creating a cached version in a provider, so the step to retrieve the data from disk every time (which will be more time consuming than from memory) can potentially be skipped. Silly to read the token from disk every time. Better to keep it in memory too.

- [x] Make version of RescadoStorage with caching

Comment (Qrivi) [reply lib/providers/api_client.dart]: Done. Was replaced by `DeviceStorage`.

**Review by Pekket — APPROVED** [review]
🧞
